build/resolver.go

```diff
@@ -0,0 +1,360 @@
+				for _, p2 := range additionalPlatforms(i, node) {
+					matcher := matcher(p2)
+					if matcher.Match(p) {
+						return i, matcher.perfect
+					}
+				}
+			}
+		}
+	}
+
+	return 0, false
+}
+
+func (r *nodeResolver) boot(ctx context.Context, idxs []int, pw progress.Writer) ([]*client.Client, error) {
+	clients := make([]*client.Client, len(idxs))
+
+	baseCtx := ctx
+	eg, ctx := errgroup.WithContext(ctx)
+
+	for i, idx := range idxs {
+		i, idx := i, idx
+		eg.Go(func() error {
+			c, err := r.clients.Do(ctx, fmt.Sprint(idx), func(ctx context.Context) (*client.Client, error) {
+				if r.nodes[idx].Driver == nil {
+					return nil, nil
+				}
+				return driver.Boot(ctx, baseCtx, r.nodes[idx].Driver, pw)
+			})
+			if err != nil {
+				return err
+			}
+			clients[i] = c
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	return clients, nil
+}
+
+func (r *nodeResolver) opts(ctx context.Context, idxs []int, pw progress.Writer) ([]gateway.BuildOpts, error) {
+	clients, err := r.boot(ctx, idxs, pw)
+	if err != nil {
+		return nil, err
+	}
+
+	bopts := make([]gateway.BuildOpts, len(clients))
+	eg, ctx := errgroup.WithContext(ctx)
+	for i, idxs := range idxs {
+		i, idx := i, idxs
+		c := clients[i]
+		if c == nil {
+			continue
+		}
+		eg.Go(func() error {
+			opt, err := r.opt.Do(ctx, fmt.Sprint(idx), func(ctx context.Context) (gateway.BuildOpts, error) {
+				opt := gateway.BuildOpts{}
+				_, err := c.Build(ctx, client.SolveOpt{
+					Internal: true,
+				}, "buildx", func(ctx context.Context, c gateway.Client) (*gateway.Result, error) {
+					opt = c.BuildOpts()
+					return nil, nil
+				}, nil)
+				return opt, err
+			})
+			if err != nil {
+				return err
+			}
+			bopts[i] = opt
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+	return bopts, nil
+}
+
+// recombineDriverPairs recombines resolved nodes that are on the same driver
+// back together into a single node.
+func recombineNodes(nodes []*resolvedNode) []*resolvedNode {
+	result := make([]*resolvedNode, 0, len(nodes))
+	lookup := map[int]int{}
+	for _, node := range nodes {
+		if idx, ok := lookup[node.driverIndex]; ok {
+			result[idx].platforms = append(result[idx].platforms, node.platforms...)
+		} else {
+			lookup[node.driverIndex] = len(result)
+			result = append(result, node)
+		}
+	}
+	return result
+}
+
+// archMatcher is an implementation of platforms.Matcher that matches only
+// against the architecture of the platform, ignoring the OS.
+type archMatcher struct {
+	platform specs.Platform
+	marcher  platforms.Matcher
+}
+
+func newArchMatcher(platform specs.Platform) *archMatcher {
+	return &archMatcher{
+		platform: platform,
+		marcher:  platforms.Only(platform),
+	}
+}
+
+func newArchMatcherStrict(platform specs.Platform) *archMatcher {
+	return &archMatcher{
+		platform: platform,
+		marcher:  platforms.OnlyStrict(platform),
+	}
+}
+
+func (m *archMatcher) Match(p specs.Platform) bool {
```

Comment:
Why would building on same arch for a different OS be faster than any other node?

Eg. if you are building for `darwin/arm64` and have nodes `linux/amd64`, `linux/arm64` then it should not go to `linux/arm64` but to the first node.

Reply:
The theory of this is not about faster speed, but about output artifacts.

OS of a system can be a bit of a lie sometimes, e.g. WSL, or Docker Desktop, etc. However, in those cases, while not perfect, it would be better to match to the architecture, instead of defaulting to just the first.

In your scenario, this could happen with a remote builder connected to Docker Desktop - if I build and `--load` without a `--platform` flag, then we'll just select the first builder, which might be the non-native architecture `linux/amd64`. We _should_ prefer to load the `linux/arm64` image to the container store instead.

It's not perfect by any measure, but I think it's a bit more reasonable than just picking the first. This way, we don't require that the user reconfigures the order of their builders (I'm not sure what the UX of that would even be like) to get the desired default behavior.

Reply:
(🤔 I thought I already left a comment for this, either I forgot to submit or it was somewhere else)

The feature you are describing here is determining the implicit platform (if none is given by user) based on `DOCKER_CONTEXT/-o context=` value if `--load` is used. If we want this then it should be specific to the `--load` behavior and look at the context value. Without `--load` it should be the native platform of the first node.

>  which might be the non-native architecture linux/amd64. 

This is only the case when there is mismatch of configurations between builder and load target. Otherwise the first is always native. I don't understand how ignoring the host allows determining the load target.

> OS of a system can be a bit of a lie sometimes, e.g. WSL, or Docker Desktop, etc

These are server-side platforms. I don't understand where the difference could come from.

Reply:
Removed this for now, I'll open a follow-up PR to do this at some point.